File: Controllers/mergeSort.py
def compareItems(itemA, itemB, reverse, flag):
    if flag == 'name' and reverse:
        return itemA.getName() > itemB.getName()
    elif flag == 'year' and reverse:
        return itemA.getYear() > itemB.getYear()
    elif flag == 'sales' and reverse:
        return itemA.getSales() > itemB.getSales()
    elif flag == 'name':
        return itemA.getName() < itemB.getName()
    elif flag == 'year':
        return itemA.getYear() < itemB.getYear()
    elif flag == 'sales':
        return itemA.getSales() < itemB.getSales()
    else:
        print('Command not found')
        return False

def mergeSort(arr, reverse, flag):
    if len(arr) > 1:
        mid = len(arr)//2
        left = arr[:mid]
        right = arr[mid:]
        mergeSort(left,reverse, flag)
        mergeSort(right,reverse, flag)
        i,j,k = (0, 0, 0)

        while i < len(left) and j < len(right):
            if compareItems(left[i], right[j], reverse, flag):
                arr[k] = left[i]
                i += 1
            else:
                arr[k] = right[j]
                j += 1

            k += 1

        while i < len(left):
            arr[k] = left[i]
            i += 1
            k += 1
 
        while j < len(right):
            arr[k] = right[j]
            j += 1
            k += 1

File: src/merge_sort.go
package main

import "github.com/projeto-de-algoritmos/D-C_GameSort/src/model"

func mergeSort(items []model.Game, isReverse bool, flag string) []model.Game {
	var num = len(items)
	if num == 1 {
		return items
	}
	middle := int(num / 2)
	var (
		left  = make([]model.Game, middle)
		right = make([]model.Game, num-middle)
	)
	for i := 0; i < num; i++ {
		if i < middle {
			left[i] = items[i]
		} else {
			right[i-middle] = items[i]
		}
	}

	return merge(mergeSort(left, isReverse, flag), mergeSort(right, isReverse, flag), isReverse, flag)
}

func merge(left, right []model.Game, isReverse bool, flag string) (result []model.Game) {
	result = make([]model.Game, len(left)+len(right))

	i := 0

	if isReverse == true && flag == "year" {
		for len(left) > 0 && len(right) > 0 {
			if left[0].GetYear() > right[0].GetYear() {
				result[i] = left[0]
				left = left[1:]
			} else {
				result[i] = right[0]
				right = right[1:]
			}
			i++
		}
	} else if isReverse == true && flag == "sales" {
		for len(left) > 0 && len(right) > 0 {
			if left[0].GetSales() > right[0].GetSales() {
				result[i] = left[0]
				left = left[1:]
			} else {
				result[i] = right[0]
				right = right[1:]
			}
			i++
		}
	} else if isReverse == true && flag == "name" {
		for len(left) > 0 && len(right) > 0 {
			if left[0].GetName() > right[0].GetName() {
				result[i] = left[0]
				left = left[1:]
			} else {
				result[i] = right[0]
				right = right[1:]
			}
			i++
		}
	} else if isReverse == false && flag == "sales" {
		for len(left) > 0 && len(right) > 0 {
			if left[0].GetSales() < right[0].GetSales() {
				result[i] = left[0]
				left = left[1:]
			} else {
				result[i] = right[0]
				right = right[1:]
			}
			i++
		}
	} else if isReverse == false && flag == "name" {
		for len(left) > 0 && len(right) > 0 {
			if left[0].GetName() < right[0].GetName() {
				result[i] = left[0]
				left = left[1:]
			} else {
				result[i] = right[0]
				right = right[1:]
			}
			i++
		}
	} else if isReverse == false && flag == "year" {
		for len(left) > 0 && len(right) > 0 {
			if left[0].GetYear() < right[0].GetYear() {
				result[i] = left[0]
				left = left[1:]
			} else {
				result[i] = right[0]
				right = right[1:]
			}
			i++
		}
	}

	for j := 0; j < len(left); j++ {
		result[i] = left[j]
		i++
	}
	for j := 0; j < len(right); j++ {
		result[i] = right[j]
		i++
	}

	return
}
